### PR TITLE
upgrade to OpenNLP 2.3.2

### PR DESCRIPTION
**Pull request by cpoerschke**
### Description

Building upon the #448 pull request upgrading to 1.9.4 -- this is a follow-up to upgrade to the latest release.

part of #11657 i.e. the _"add support for the new interface implementations in the OpenNLP analysis module that was added in #3973"_ part is not included in this pull request

### CODE REVIEW

File: lucene/analysis/opennlp/src/test/org/apache/lucene/analysis/opennlp/TestOpenNLPChunkerFilterFactory.java [hunk 1]
```diff
@@ -58,7 +58,7 @@ public class TestOpenNLPChunkerFilterFactory extends BaseTokenStreamTestCase {
     8, 15, 17, 21, 23, 29, 30, 39, 46, 48, 49, 51, 57, 58
   };
   private static final String[] SENTENCES_chunks = {
-    "B-NP", "I-NP", "I-NP", "B-VP", "B-NP", "I-NP", "O", "B-NP", "I-NP", "I-NP", "O", "B-NP",
+    "B-NP", "I-NP", "I-NP", "I-NP", "I-NP", "I-NP", "O", "B-NP", "I-NP", "I-NP", "O", "B-NP",
```
Comment (cpoerschke) [inline lucene/analysis/opennlp/src/test/org/apache/lucene/analysis/opennlp/TestOpenNLPChunkerFilterFactory.java:61]: This change so far is naively what makes the tests pass ... however

```suggestion
    "B-NP", "I-NP", "I-NP", "I-VP", "I-NP", "I-NP", "O", "B-NP", "I-NP", "I-NP", "O", "B-NP",
```

i.e. `"I-VP"` for `"has"` would seem more correct, no?

File: lucene/licenses/opennlp-tools-NOTICE.txt [hunk 2]
```diff
@@ -1,6 +1,11 @@
...
+Copyright 2017 The Apache Software Foundation
 
 This product includes software developed at
 The Apache Software Foundation (http://www.apache.org/).
+
+The snowball stemmers in
+opennlp-tools/src/main/java/opennlp/tools/stemmer/snowball
+were developed by Martin Porter and Richard Boulton.
+The full snowball package is available from
+http://snowball.tartarus.org/
```
Comment (cpoerschke) [inline lucene/licenses/opennlp-tools-NOTICE.txt:11]: TODO: https://github.com/apache/opennlp/blob/opennlp-2.3.0/NOTICE

File: lucene/analysis/opennlp/src/tools/test-model-data/README.txt [hunk 3]
```diff
@@ -3,4 +3,4 @@ Training data derived from Reuters corpus in very unscientific way.
 Tagging done with CCG Urbana-Champaign online demos:
 	http://cogcomp.cs.illinois.edu/page/demos
 
-Run 'ant train-test-models' to generate models from training data here.
```
Comment (cpoerschke) [inline lucene/analysis/opennlp/src/tools/test-model-data/README.txt:6]: This is where the `ant train-test-models` target is defined: https://github.com/apache/lucene-solr/blob/releases/lucene-solr/8.11.2/lucene/analysis/opennlp/build.xml#L52-L84

**Comment by epugh** [comment]
FYI 2.3.1 was just released.

File: lucene/analysis/opennlp/src/test/org/apache/lucene/analysis/opennlp/TestOpenNLPChunkerFilterFactory.java (hunk 1 above)
Comment (cpoerschke) [reply lucene/analysis/opennlp/src/test/org/apache/lucene/analysis/opennlp/TestOpenNLPChunkerFilterFactory.java:61]: > ... would seem more correct, no?

Upon further consideration, the _"Tagging models are created from tiny test data in opennlp/tools/test-model-data/ and are not very accurate."_ comment from line 31-32 above applies here probably.

File: lucene/licenses/slf4j-api-LICENSE-MIT.txt [hunk 4]
```diff
@@ -0,0 +1,24 @@
+Copyright (c) 2004-2022 QOS.ch Sarl (Switzerland)
```
Comment (cpoerschke) [inline lucene/licenses/slf4j-api-LICENSE-MIT.txt:1]: https://github.com/qos-ch/slf4j/blob/v_1.7.36/slf4j-api/LICENSE.txt

**Comment by dweiss** [comment]
Hi @cpoerschke - apologies for the delay. This looks ok to me. It is a pity the dependency sucks in so much other software (slf4j included). I guess this is unavoidable if somebody wants to use it though.

It would be good migrate the model regeneration to gradle - probably a good follow-up issue on its own though.

**Review by dweiss — COMMENTED** [review]
What's probably missing is the changes entry and a note on migration?

File: lucene/analysis/opennlp/src/tools/test-model-data/README.txt (hunk 3 above)
Comment (cpoerschke) [reply lucene/analysis/opennlp/src/tools/test-model-data/README.txt:6]: #13002 is for migration to Gradle.

**Comment by cpoerschke** [comment]
> What's probably missing is ... a note on migration?

Yes, indeed. It being a 1.x to 2.x dependency upgrade, I'm not yet familiar enough with OpenNLP to gauge what's involved in migrating. Perhaps @jzonthemtn and/or @epugh would have insights to share on that? Thanks!

**Comment by epugh** [comment]
I don't know about the migration side specifically, but based on https://opennlp.apache.org/news/release-200.html and some of the other release notes, here is a first stab.  

```
OpenNLP 2x opens the door to accessing various models via the ONNX runtime.  To migrate you will need to update any deprecated OpenNLP methods that you may be using and be running on Java 17.
```

**Comment by cpoerschke** [comment]
> FYI 2.3.1 was just released.

2.3.2 just released -- https://opennlp.apache.org/news/release-232.html -- I'll update this PR

**Review by epugh — APPROVED** [review]
LGTM!   Thanks for getting this plumbing in, if we get it in the next version of Lucene, then Solr 9.6 can take advantage of this ONNX stuff without weird gradle stuff!

**Review by dweiss — COMMENTED** [review]
I will take a look at this, promise. It is a bit disappointing that the upgrade pulls in slf4j dependency... I'm also surprised there are no other transitive dependencies in the lock file (the notice file hints at their presence?).

File: lucene/licenses/opennlp-tools-NOTICE.txt [hunk 5]
```diff
@@ -1,11 +1,101 @@
...
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+
+============================================================================
+
+ONNX Runtime
```
Comment (dweiss) [inline lucene/licenses/opennlp-tools-NOTICE.txt:44]: Interestingly, only opennlp-dl has onnx links - the opennlp-tools doesn't. Maybe we should remove this (can we?). I see opennlp notice file includes it... but still - wouldn't it be confusing to Lucene users willing to read through the legal stuff?

File: lucene/licenses/opennlp-tools-NOTICE.txt [hunk 6]
```diff
@@ -1,11 +1,101 @@
...
+EXPRESS OR  IMPLIED, INCLUDING  BUT NOT LIMITED  TO THE  WARRANTIES OF
+MERCHANTABILITY,    FITNESS    FOR    A   PARTICULAR    PURPOSE    AND
+NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+OF CONTRACT, TORT OR OTHERWISE,  ARISING FROM, OUT OF OR IN CONNECTION
+WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+============================================================================
+
+jackson-databind
```
Comment (dweiss) [inline lucene/licenses/opennlp-tools-NOTICE.txt:99]: Same concern here (not used by us or opennlp-tools).

File: versions.lock [hunk 7]
```diff
@@ -20,6 +20,7 @@ org.openjdk.jmh:jmh-core:1.37 (1 constraints: df04fc30)
 org.ow2.asm:asm:9.6 (3 constraints: 3917ef6d)
 org.ow2.asm:asm-commons:9.6 (1 constraints: b304382c)
 org.ow2.asm:asm-tree:9.6 (1 constraints: ea09e3a5)
+org.slf4j:slf4j-api:1.7.36 (1 constraints: 6f0ed053)
```
Comment (dweiss) [inline versions.lock:23]: There is no way to dodge this dependency, sadly.

**Review by dweiss — APPROVED** [review]
I think this looks good. The only concern for me is about including notice mentions of stuff we're not actually using or referencing. I think this is confusing. We can merge this in and perhaps fire an email to legal at apache, asking whether the notice file of opennlp can be trimmed if one is not using all of their modules?

**Comment by cpoerschke** [comment]
> ... perhaps fire an email to legal at apache, asking ...

Good idea, will do. And thanks for reviewing also with this in mind!

**Comment by uschindler** [comment]
Hi @cpoerschke,
the backport to 9.x produces errors, as opennlp seems to require Java 17:

```
/home/jenkins/workspace/Lucene-9.x-Linux/lucene/analysis/opennlp/src/java/org/apache/lucene/analysis/opennlp/tools/NLPChunkerOp.java:21: error: cannot access ChunkerME
import opennlp.tools.chunker.ChunkerME;
                            ^
  bad class file: /home/jenkins/.gradle/caches/modules-2/files-2.1/org.apache.opennlp/opennlp-tools/2.3.2/d739edba1e729691ed5ab80e1ccf330555a02ea7/opennlp-tools-2.3.2.jar(/opennlp/tools/chunker/ChunkerME.class)
    class file has wrong version 61.0, should be 55.0
    Please remove or make sure it appears in the correct subdirectory of the classpath.
```

I think we have to revert the upgrade in Lucene 9.x and only apply it to main (Lucene 10). It should not take too long until we release 10.0, so I think that's a fine compromise.

If you revert the commit on 9.x, please move the CHANGES.txt entry in main branch.

Uwe

**Comment by uschindler** [comment]
Here is the confirmation: OpenNLP 2.3 raised minimum Java version to 17: https://opennlp.apache.org/news/release-230.html

So we need to revert this on 9.x branch.

**Comment by dweiss** [comment]
@cpoerschke I think we need to revert this commit from 9x because the failed builds keep polluting the mailing list. I know it's the weekend, so I'll do this without waiting for your permission - hope you're ok with this while we're looking into alternatives.

**Comment by uschindler** [comment]
OpenNLP 2.2 would be an option, which ist still compatible to Java 11.
But I don't think we should spend too much time with this. Let's release 10.0 at some point.

**Comment by uschindler** [comment]
This PR also added a migration guide entry. Maybe move it, too.

**Comment by dweiss** [comment]
I have, see commit 7d99e6aafbdb32e1a6c94e10e41d58a1f790630a

**Comment by cpoerschke** [comment]
Thanks @uschindler and @dweiss for reverting on `branch_9x` here!

minor: pushed https://github.com/apache/lucene/commit/8bf10130e95c7b1b3da16606be0bf872c28fb6ed to remove the empty 9.9 to 9.10 migrate section.

**Comment by jzonthemtn** [comment]
Thanks everyone involved in this!

**Pull request merged**